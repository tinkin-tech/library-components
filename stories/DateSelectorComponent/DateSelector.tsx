import * as React from 'react'

import DateUtils from '../../utils/dateUtils'
import { LabelComponent } from '../label/Label'

type FormatDateTypes = 'YYYY-MM-DD'

interface PropsInterface {
  id: string
  date: string
  onChangeDate: (id: string, date: string) => void
  dateFormat: FormatDateTypes
  label?: string
  error?: string
  required?: boolean
  minDate?: string
  maxDate?: string
  disable?: boolean
}

export const DateSelector = (
  props: PropsInterface
): React.ReactElement<PropsInterface> => {
  const {
    onChangeDate,
    dateFormat,
    id,
    date,
    minDate,
    maxDate,
    label,
    disable,
    required,
    error,
  } = props
  const [separateDate, changeSeparateDate] = React.useState({
    year: '',
    month: '',
    day: '',
  })
  const [selectedValuesState, changeSelectedValuesState] = React.useState([])
  const [dateState, changeDateState] = React.useState(date)
  const [openSelectors, changeOpenSelectors] = React.useState({
    year: false,
    month: false,
    day: false,
  })
  const onChangeInput = (
    e?: React.FormEvent<HTMLInputElement>,
    idInput?: string,
    valueInput?: string
  ): void => {
    let dateToChange = { year: '', month: '', day: '' }
    const dateAdded = DateUtils.addDate(
      minDate || dateState || date,
      dateFormat,
      1,
      'days'
    )
    const dateAddedTransformed = DateUtils.transformDateStringToDate(
      dateAdded || dateState || date,
      dateFormat
    )
    switch (e?.currentTarget?.id || idInput) {
      case 'year':
        dateToChange = {
          year: e?.currentTarget?.value || valueInput,
          month: minDate
            ? dateAddedTransformed.getMonth() + 1 >= 10
              ? (dateAddedTransformed.getMonth() + 1).toString()
              : `0${(dateAddedTransformed.getMonth() + 1).toString()}`
            : '01',
          day: minDate
            ? (dateAddedTransformed.getDate() >= 10
                ? dateAddedTransformed.getDate()
                : `0${dateAddedTransformed.getDate()}`
              ).toString()
            : '01',
        }
        changeSeparateDate(dateToChange)
        break
      case 'month':
        dateToChange = {
          ...separateDate,
          month: e?.currentTarget?.value || valueInput,
          day: minDate
            ? (dateAddedTransformed.getDate() >= 10
                ? dateAddedTransformed.getDate()
                : `0${dateAddedTransformed.getDate()}`
              ).toString()
            : '01',
        }
        changeSeparateDate(dateToChange)
        break
      case 'day':
        dateToChange = {
          ...separateDate,
          day: e?.currentTarget?.value || valueInput,
        }
        changeSeparateDate(dateToChange)
        break
    }
    changeDateState(
      `${dateToChange.year}-${dateToChange.month}-${dateToChange.day}`
    )
    onChangeDate(
      id,
      `${dateToChange.year}-${dateToChange.month}-${dateToChange.day}`
    )
  }

  const getSelectValues = (e: React.FormEvent<HTMLInputElement>): void => {
    const selectedValues: string[] = []
    const firstDate =
      maxDate || DateUtils.addDate(date || dateState, dateFormat, 5, 'years')
    switch (e.currentTarget.id) {
      case 'year':
        const secondDate =
          minDate ||
          DateUtils.substractDate(date || dateState, dateFormat, 5, 'years')
        new Array(DateUtils.yearsDiff(firstDate, secondDate, dateFormat) + 1)
          .fill(0)
          .forEach((_, index) => {
            const year = DateUtils.getFullYear(secondDate, dateFormat)
            selectedValues.push((year + index).toString())
          })
        changeOpenSelectors({
          year: !openSelectors.year,
          month: false,
          day: false,
        })
        break
      case 'month':
        const monthsDiff = DateUtils.monthsDiff(
          firstDate,
          dateState || date,
          dateFormat
        )
        const months = monthsDiff + 1 > 12 ? 12 : monthsDiff + 1
        new Array(months).fill(0).forEach((item, index) => {
          const month = item + 1 + index
          selectedValues.push(month >= 10 ? month.toString() : `0${month}`)
        })
        changeOpenSelectors({
          year: false,
          month: !openSelectors.month,
          day: false,
        })
        break
      case 'day':
        const dateForCompare = DateUtils.setDay(
          dateState || date,
          dateFormat,
          1
        )
        const daysInMonth = DateUtils.getDaysInMonth(
          dateState || date,
          dateFormat
        )
        const daysDiff = DateUtils.daysDiff(
          firstDate,
          dateForCompare,
          dateFormat
        )
        const days = daysDiff + 1 >= daysInMonth ? daysInMonth : daysDiff + 1
        new Array(days).fill(0).forEach((item, index) => {
          const day = item + 1 + index
          selectedValues.push(day >= 10 ? day.toString() : `0${day}`)
        })
        changeOpenSelectors({
          year: false,
          month: false,
          day: !openSelectors.day,
        })
    }
    changeSelectedValuesState(selectedValues)
  }

  const selectorComponent = (
    testId: string,
    inputId: string
  ): React.ReactElement => (
    <div className="selector app-scroll shadow-2" data-testid={testId}>
      {selectedValuesState.map((value, index) => {
        const handleOnClickSelectItem = (): void => {
          onChangeInput(null, inputId, value)
          changeOpenSelectors({ ...openSelectors, [inputId]: false })
        }
        return (
          <div
            key={index}
            className="select-items"
            data-testid={`${testId}-item-${index}`}
            onClick={handleOnClickSelectItem}
          >
            {value}
          </div>
        )
      })}
    </div>
  )

  React.useEffect(() => {
    const splitDate = dateState ? dateState.split('-') : date.split('-')
    changeSeparateDate({
      year: splitDate[0] || '',
      month: splitDate[1] || '',
      day: splitDate[2] || '',
    })
  }, [date])

  const selectOptions = React.useRef()

  React.useEffect(() => {
    const eventListener = (e: Event): void => {
      if (
        e.target !== selectOptions.current &&
        (openSelectors.day || openSelectors.month || openSelectors.year)
      ) {
        changeOpenSelectors({ day: false, month: false, year: false })
      }
    }
    window.addEventListener('click', eventListener)
    return (): void => window.removeEventListener('click', eventListener)
  })

  return (
    <div
      className={`dateSelectorComponent`}
      data-testid="dateSelectorComponent"
      ref={selectOptions}
    >
      <div className="flex-space-between">
        <LabelComponent
          label={label}
          required={required}
          error={!!error}
          referenceId={id}
        />
        {error && <span className="icon-warning-content">&#9888;</span>}
      </div>
      <div className="dateSelectorContent">
        <div className="dateContainer">
          <div className="dateContent flex-row flex-middle">
            <span className="inputContent flex-column flex-1">
              <input
                data-testid="input-year"
                onChange={onChangeInput}
                id="year"
                value={separateDate.year}
                onClick={getSelectValues}
                disabled={disable}
                placeholder="YYYY"
              />
            </span>
            <span className="iconContent p-r flex-column">
              <i
                className={`arrow down ${openSelectors.year ? 'active' : ''}`}
              />
            </span>
          </div>
          {openSelectors.year && selectorComponent('selector-year', 'year')}
        </div>
        <div className="dateContainer">
          <input
            data-testid="input-month"
            onChange={onChangeInput}
            id="month"
            value={separateDate.month}
            disabled={!separateDate.year || disable}
            onClick={getSelectValues}
            placeholder="MM"
          />
          {openSelectors.month && selectorComponent('selector-month', 'month')}
        </div>
        <div className="dateContainer">
          <input
            data-testid="input-day"
            onChange={onChangeInput}
            id="day"
            value={separateDate.day}
            disabled={!separateDate.month || disable}
            onClick={getSelectValues}
            placeholder="DD"
          />
          {openSelectors.day && selectorComponent('selector-day', 'day')}
        </div>
      </div>
      {error && <span className="error">{error}</span>}
    </div>
  )
}
