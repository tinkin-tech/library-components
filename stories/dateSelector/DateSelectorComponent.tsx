import * as React from 'react'
import DateUtils from '../../utils/dateUtils'

type FormatDateTypes = 'YYYY-MM-DD'

interface PropsInterface {
  id: string
  date: string
  onChangeDate: (id: string, date: string) => void
  dateFormat: FormatDateTypes
  label?: string
  error?: string
  required?: boolean
  minDate?: string
  maxDate?: string
}

const DateSelectorComponent = (
  props: PropsInterface
): React.ReactElement<PropsInterface> => {
  const { onChangeDate, id, date } = props
  const [separateDate, changeSeparateDate] = React.useState({
    year: '',
    month: '',
    day: '',
  })
  const onChangeInput = (e: React.FormEvent<HTMLInputElement>): void => {
    let dateToChange = { year: '', month: '', day: '' }
    switch (e.currentTarget.id) {
      case 'year':
        dateToChange = { year: e.currentTarget.value, month: '', day: '' }
        changeSeparateDate(dateToChange)
        break
      case 'month':
        dateToChange = {
          ...separateDate,
          month: e.currentTarget.value,
          day: '',
        }
        changeSeparateDate(dateToChange)
        break
      case 'day':
        dateToChange = { ...separateDate, day: e.currentTarget.value }
        changeSeparateDate(dateToChange)
        break
    }
    onChangeDate(
      id,
      `${dateToChange.year}-${dateToChange.month}-${dateToChange.day}`
    )
  }

  React.useEffect(() => {
    const splitDate = date.split('-')
    changeSeparateDate({
      year: splitDate[0] || '',
      month: splitDate[1] || '',
      day: splitDate[2] || '',
    })
  }, [date])

  return (
    <div data-testid="dateSelectorComponent">
      otnuehut
      <div>oeueoueo</div>
      <input
        data-testid="input-year"
        onChange={onChangeInput}
        id="year"
        value={separateDate.year}
      />
      <input
        data-testid="input-month"
        onChange={onChangeInput}
        id="month"
        value={separateDate.month}
        disabled={!separateDate.year}
      />
      <input
        data-testid="input-day"
        onChange={onChangeInput}
        id="day"
        value={separateDate.day}
        disabled={!separateDate.month}
      />
    </div>
  )
}

export default React.memo(DateSelectorComponent)
