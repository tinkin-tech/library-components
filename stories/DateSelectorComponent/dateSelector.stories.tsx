import * as React from 'react'
import { DateSelector } from './DateSelector'

export default {
  title: 'Date Selector',
  component: DateSelector,
}

type formatType = 'YYYY-MM-DD'

interface PropsInterface {
  id: string
  date: string
  dateFormat: formatType
  onChangeDate: (id: string, value: string) => void
}

const Template = (args: PropsInterface): React.ReactElement => (
  <DateSelector {...args} />
)

export const WithValue = Template.bind({})

WithValue.args = {
  id: 'DateSelectorComponent',
  date: '2020-01-01',
  minDate: '2020-10-11',
  maxDate: '2020-12-25',
  dateFormat: 'YYYY-MM-DD',
  onChangeDate: (): void => null,
}
