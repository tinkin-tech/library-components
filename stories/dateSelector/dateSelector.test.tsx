import * as React from 'react'
import * as renderer from 'react-test-renderer'
import DateSelectorComponent from './DateSelectorComponent'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

const dateSelectorDefault = (
  <DateSelectorComponent
    date="20-07-2020"
    dateFormat="YYYY-MM-DD"
    label="date selector"
    onChangeDate={(): null => null}
    error=""
    required={true}
    minDate="2020-07-18"
    maxDate="2020-07-22"
    id="dateSelector"
  />
)

describe('', () => {
  it('Should render in virtual DOM', () => {
    const { getByTestId } = render(dateSelectorDefault)
    expect(getByTestId('dateSelectorComponent')).toBeInTheDocument()
  })

  it('Should shot the snapshot with all properties', () => {
    const component = renderer.create(dateSelectorDefault)
    const componentJSON = component.toJSON()
    expect(componentJSON).toMatchSnapshot()
  })

  it(
    'Should onChangeDate function return id passed in properties and date' +
      'change year when change year input',
    () => {
      const onChangeFake = jest.fn((id: string, date: string): {
        id: string
        date: string
      } => ({ id, date }))
      const { getByTestId } = render(
        <DateSelectorComponent
          id="dateSelector"
          date="2020-07-20"
          onChangeDate={onChangeFake}
          dateFormat="YYYY-MM-DD"
        />
      )
      renderer.act(() => {
        fireEvent.change(getByTestId('input-year'), {
          target: { value: '2018' },
        })
      })
      expect(onChangeFake).toHaveBeenCalledTimes(1)
      expect(onChangeFake.mock.results[0].value).toMatchObject({
        id: 'dateSelector',
        date: '2018--',
      })
    }
  )

  it(
    'Should onChangeDate function return id passed in properties and date ' +
      'changed month when change month input',
    () => {
      const onChangeFake = jest.fn((id: string, date: string): {
        id: string
        date: string
      } => ({ id, date }))
      const { getByTestId } = render(
        <DateSelectorComponent
          id="selector"
          date="2019-12-10"
          onChangeDate={onChangeFake}
          dateFormat="YYYY-MM-DD"
        />
      )
      renderer.act(() => {
        fireEvent.change(getByTestId('input-month'), {
          target: { value: '03' },
        })
      })
      expect(onChangeFake.mock.results[0].value).toMatchObject({
        id: 'selector',
        date: '2019-03-',
      })
    }
  )

  it(
    'Should onChangeDate function return id passed in properties and date' +
      'changed day when change day input',
    () => {
      const onChangeFake = jest.fn((id: string, date: string): {
        id: string
        date: string
      } => ({ id, date }))
      const { getByTestId } = render(
        <DateSelectorComponent
          id="selector"
          date="2020-07-22"
          onChangeDate={onChangeFake}
          dateFormat="YYYY-MM-DD"
        />
      )
      renderer.act(() => {
        fireEvent.change(getByTestId('input-day'), {
          target: { value: '11' },
        })
      })
      expect(onChangeFake.mock.results[0].value).toMatchObject({
        id: 'selector',
        date: '2020-07-11',
      })
    }
  )

  it('Should set month and day empty when change year input', () => {
    const onChangeFake = jest.fn((id: string, date: string): {
      id: string
      date: string
    } => ({ id, date }))
    const { getByTestId } = render(
      <DateSelectorComponent
        id="selector"
        date="2020-07-22"
        onChangeDate={onChangeFake}
        dateFormat="YYYY-MM-DD"
      />
    )
    renderer.act(() => {
      fireEvent.change(getByTestId('input-year'), {
        target: { value: '2010' },
      })
    })
    expect(onChangeFake.mock.results[0].value).toMatchObject({
      id: 'selector',
      date: '2010--',
    })
  })

  it('Should set day empty when change month input', () => {
    const onChangeFake = jest.fn((id: string, date: string): {
      id: string
      date: string
    } => ({ id, date }))
    const { getByTestId } = render(
      <DateSelectorComponent
        id="selector"
        date="2020-07-22"
        onChangeDate={onChangeFake}
        dateFormat="YYYY-MM-DD"
      />
    )
    renderer.act(() => {
      fireEvent.change(getByTestId('input-month'), {
        target: { value: '02' },
      })
    })
    expect(onChangeFake.mock.results[0].value).toMatchObject({
      id: 'selector',
      date: '2020-02-',
    })
  })

  it('Should set empty string in input values when pass empty date', () => {
    const { getByTestId } = render(
      <DateSelectorComponent
        id="datepicker"
        date=""
        onChangeDate={null}
        dateFormat="YYYY-MM-DD"
      />
    )
    const yearInput = getByTestId('input-year')
    const monthInput = getByTestId('input-month')
    const dayInput = getByTestId('input-day')
    expect(yearInput).toMatchObject({ value: '' })
    expect(monthInput).toMatchObject({ value: '' })
    expect(dayInput).toMatchObject({ value: '' })
  })

  it(
    'Should month input have disable property when have date without year ' +
      'in props',
    () => {
      const onChangeFake = jest.fn((id: string, date: string): {
        id: string
        date: string
      } => ({ id, date }))
      const { getByTestId } = render(
        <DateSelectorComponent
          id="date"
          date=""
          onChangeDate={onChangeFake}
          dateFormat="YYYY-MM-DD"
        />
      )
      const inputMonth = getByTestId('input-month')
      expect(inputMonth).toMatchObject({ disabled: true })
    }
  )

  it(
    'Should day input have disable property when have date without month' +
      'in props',
    () => {
      const onChangeFake = jest.fn((id: string, date: string): {
        id: string
        date: string
      } => ({ id, date }))
      const { getByTestId } = render(
        <DateSelectorComponent
          id="date"
          date="2020"
          onChangeDate={onChangeFake}
          dateFormat="YYYY-MM-DD"
        />
      )
      const inputDay = getByTestId('input-day')
      expect(inputDay).toMatchObject({ disabled: true })
    }
  )

  it(
    'Should month input have disable property in false when have date' +
      ' with year',
    () => {
      const onChangeFake = jest.fn((id: string, date: string): {
        id: string
        date: string
      } => ({ id, date }))
      const { getByTestId } = render(
        <DateSelectorComponent
          id="date"
          date="2020"
          onChangeDate={onChangeFake}
          dateFormat="YYYY-MM-DD"
        />
      )
      const inputMonth = getByTestId('input-month')
      expect(inputMonth).toMatchObject({ disabled: false })
    }
  )
})
