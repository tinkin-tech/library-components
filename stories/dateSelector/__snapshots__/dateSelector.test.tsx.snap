// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should shot the snapshot with all properties 1`] = `
<div
  data-testid="dateSelectorComponent"
>
  otnuehut
  <div>
    oeueoueo
  </div>
  <input
    data-testid="input-year"
    id="year"
    onChange={[Function]}
    value=""
  />
  <input
    data-testid="input-month"
    disabled={true}
    id="month"
    onChange={[Function]}
    value=""
  />
  <input
    data-testid="input-day"
    disabled={true}
    id="day"
    onChange={[Function]}
    value=""
  />
</div>
`;
